pkg/services/vimmachine.go

```diff
@@ -390,7 +390,10 @@ func (v *VimMachineService) createOrPatchVSphereVM(ctx *context.VIMMachineContex
 		if vm.Spec.Server == "" {
 			vm.Spec.Server = ctx.VSphereCluster.Spec.Server
 		}
-		if vm.Spec.Thumbprint == "" {
+		// Need to cover two cases:
+		// 1. vm.Spec.Thumbprint is empty after copy from VSphereMachine.Spec
+		// 2. VSphereCluster.Spec.Thumbprint is updated by user, such as when the certificates of the vCenter instance are updated
+		if vm.Spec.Thumbprint == "" || vm.Spec.Thumbprint != ctx.VSphereCluster.Spec.Thumbprint {
```

Comment:
IMO this change is not necessary, but am open to listen to your thoughts here.
A bit of background, the `mutateFn` defined on line 340 aims to recreate a new copy of the `VSphereVM` object using the same values it originally used to create the `VSphereVM` object. This object is then passed to `patchOrCreate` which (in this case) updates the object stored in the API server.
So the check (on line 393, referring to the original file before the PR) is going to be true if the `VSphereMachine.Spec.thumbprint` is empty. If that is the case, then copy the value from the `VSphereCluster.Spec.thumbprint` into the VSphereVM object. 
Once the call to `patchOrCreate` is finished, you will see the VSphereVM object get updated with the newer thumbprint value.

Reply:
Hi @srm09,
I'm totally agree with you about the case you mentioned, VSphereVM thumbprint is empty here so everything works fine with original logic.

I have a concern for another case, that user also specified thumbprint in VSphereMachineTemplate, so `VSphereMachine.Spec.thumbprint` has an old thumbprint, when user want to update VSphereCluster and VSphereVM with new thumbprint, according to [Line 337 ](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/pkg/services/vimmachine.go#L377) , `vm.Spec.Thumbprint` will deep copy from `VSphereMachine.Spec.thumbprint` with old thumbprint, so it is not empty and the check is fail, I'm afraid it never get the chance to update with new thumbprint. What do you think?

Reply:
For the case when the user is consuming the thumbprint via the VSphereMachineTemplate, the way to move forward with the update is to create a new template with the updated thumbprint value and use that for machine creation.

Reply:
This PR should handle the case when thumbprint is not set on the `VSphereMachine.Spec`, and is consumed by the VSphereCluster object.

Reply:
As if `VSphereMachine.Spec.thumbprint` is not empty, user shouldn't use this approach (updating `VSphereCluster`)to update VSpherevm.Spec.Thumbprint, so original logic works fine. I'll revert the change. Thanks for the clarification!